packages/legend-query-builder/src/components/QueryBuilder.tsx

```diff
@@ -367,10 +371,70 @@ export const QueryBuilder = observer(
       }
       return null;
     };
+
+    const undo = useCallback((): void => {
+      queryBuilderState.changeHistoryState.undo();
+    }, [queryBuilderState.changeHistoryState]);
+
+    const redo = useCallback((): void => {
+      queryBuilderState.changeHistoryState.redo();
+    }, [queryBuilderState.changeHistoryState]);
+
+    useEffect(() => {
+      // this condition is for passing all exisitng tests because when we initialize a queryBuilderState for a test,
+      // we use an empty RawLambda with an empty class and this useEffect is called earlier than initializeWithQuery()
+      if (
+        (queryBuilderState.isQuerySupported && queryBuilderState.class) ||
+        !queryBuilderState.isQuerySupported
+      ) {
+        queryBuilderState.changeHistoryState.cacheNewQuery(
+          queryBuilderState.buildQuery(),
+        );
+      }
+    }, [queryBuilderState, queryBuilderState.hashCode]);
+
+    // bind ctl + z to undo, ctl + y to redo
+    useEffect(() => {
+      const onCtrlZ = (event: KeyboardEvent): void => {
+        if (
+          event.ctrlKey &&
+          event.key === 'z' &&
+          // make undo/redo is contextual e.g. if there is a new modal open e.g. parameter modal, it won't close this modal and modify underlying query
+          queryBuilderRef.current &&
+          (queryBuilderRef.current === document.activeElement ||
```

Comment:
May be you could check if parameters/constants/watermark/other modals are not opened and then perform operations in the undo() and redo() code that way you will be able to leverage registerCommands feature we have